### ISSUE TITLE
how do I undo paste?

### ISSUE THREAD

**Issue by ri-aje** [ranger/ranger#2037]
#### Runtime Environment

- Operating system and version: 
- Terminal emulator and version: 
- Python version: 
- Ranger version/commit: 
ranger version: ranger-master v1.9.3-27-gd17e532a
Python version: 3.7.1 (default, Sep 21 2019, 10:53:58) [GCC 7.4.0]
Locale: en_US.UTF-8
- Locale: 

#### Current Behavior
under the `u?` menu, I didn't find an undo paste command, looks like not doable.

#### Expected Behavior
would like to nullify the effect of a wrong paste. 

#### Context

this happened to me when I pasted some files from a subdir into its parent dir, only to find the same named files already exist. ranger was nice enough to not overwrite the targets, but instead renamed the paste extension and have them live side by side. but this does mess up the view as each `*.ext` is now followed by a `*.ext_` which broke the continuity between the original `*.ext` files. would like an undo paste feature to nicely put pasted files back to where they were.

currently I have to filter `ext_` to focus on the pasted files, then select and cut them, then navigate to their origin and paste them back there. and this hasn't considered the messy cases where not all pasted files have name collision. it's not hard to imagine some files got conflicted and renamed, while the rest just pasted fine. now picking the pasted ones out is much harder. it'll be wonderful if ranger remembers the paste operation and be able to undo it on one click.

#### Possible Solutions
keep the cp/mv commands for last paste operation and reverse the operands on undo? 

#### Steps to reproduce
1. 
2. 
3. 

#### Traceback

```

```

**Comment by toonn** [comment]
This would involve some pretty major changes. Ranger would have to keep a history and there's no guarantee it can undo a move. Imagine moving `A/a` to `B/`, creating a new `A/a` then trying to undo the move.

Re-selecting whatever files were pasted last would probably be easier and have less corner cases.

**Comment by ri-aje** [comment]
> This would involve some pretty major changes. Ranger would have to keep a history and there's no guarantee it can undo a move. Imagine moving `A/a` to `B/`, creating a new `A/a` then trying to undo the move.
> 
> Re-selecting whatever files were pasted last would probably be easier and have less corner cases.

one key re-selecting is closer to what I am after, if the north star is too complicated. how do I re-select?

**Comment by toonn** [comment]
*Implementing* being able to re-select pasted files would be easier.

**Comment by ri-aje** [comment]
> _Implementing_ being able to re-select pasted files would be easier.

understand it's not available today. my question was when do you think ranger will have it? thanks.

**Comment by toonn** [comment]
It's not a high priority for me tbh. The quickest way would be for you to implement it : )

**Comment by ri-aje** [comment]
> It's not a high priority for me tbh. The quickest way would be for you to implement it : )

ok. where to start looking?

**Comment by toonn** [comment]
Probably familiarize yourself with the copy buffer. I'd grep for "copy_buffer" or "copybuffer."

**Comment by kevinhwang91** [comment]
In my option, I think it's massive work to implement this feature. Not only  for undo but also redo and the history of undo.
You need to parse and write a rollback log for cut, bulkname, rename, and so on, which is similar to SQLite.

**Comment by toonn** [comment]
Yes, I agree rollback would require a lot of work. I think @ri-aje was going to try and implement re-selecting whatever was pasted last though, which is a lot easier. Of course, if you paste with `overwrite=True` the overwritten files are lost.

**Comment by ri-aje** [comment]
@toonn 
I made an impl for re-selecting last pasted items here https://github.com/ranger/ranger/compare/master...ri-aje:mark.last.pasted. `mark_paste` only re-selects those in current dir. manually tested copy/cut -> paste with some common cases including soft links, seems working as expected.

**Comment by sebhahn** [comment]
I came across a similar issue moving/pasting the wrong files, mainly because unwanted files (e.g. from a previous selection) have not been removed from the cut list. Is it somehow possible to view all files which are currently in the list of selected files to be moved/pasted?

**Comment by toonn** [comment]
@sebhahn, not really. I think the easiest way to do it is `:shell echo %c` and then `S` to see the result. Or you can use `:save_copy_buffer` and view the `$XDG_CONFIG_HOME/ranger/copy_buffer` file. Feel free to open a feature request.

**Comment by galaxy-gao** [comment]
> @toonn I made an impl for re-selecting last pasted items here [master...ri-aje:mark.last.pasted](https://github.com/ranger/ranger/compare/master...ri-aje:mark.last.pasted). `mark_paste` only re-selects those in current dir. manually tested copy/cut -> paste with some common cases including soft links, seems working as expected.

@ri-aje could you create a pull request for this feature?
I need it also.
thanks.

**Comment by galaxy-gao** [comment]
> > @ri-aje could you create a pull request for this feature?
> > I need it also.
> > thanks.
> 
> I shared my changes above.

ok, I got it.
I just think it would be better to merge this feature to master, or make it a plugin if possible.
